File: stack/decodeString.py
# Decode String

s="3[a]2[bc]"

def decodeString(s):
    st=[]
    for c in s:
        if c!=']':
            st.append(c)
        else:
            substr=""
            while st[-1]!='[':
                substr=st.pop()+substr
            st.pop()
            num=""
            while st and st[-1].isdigit():
                num=st.pop()+num
            st.append(int(num)*substr)
    return "".join(st)

print(decodeString(s)) # aaabcbc

File: stack/decodeString.java
package stack;
import java.util.*;
public class decodeString {
    public static void main(String[] args) {
        String s = "3[a2[c]]";
        System.out.println(decodestring(s));
    }

    public static String decodestring(String s){
        Stack<Character> st = new Stack<>();
        for (char ch : s.toCharArray()) {
            if (ch != ']') {
                st.push(ch);
            } else {
                StringBuilder sb = new StringBuilder();
                while (!st.isEmpty() && st.peek() != '[') {
                    sb.insert(0, st.pop());
                }
                st.pop();
                StringBuilder num = new StringBuilder();
                while (!st.isEmpty() && Character.isDigit(st.peek())) {
                    num.insert(0, st.pop());
                }
                int k = Integer.parseInt(num.toString());
                String str = sb.toString();
                while (k-- > 0) {
                    for (char c : str.toCharArray()) {
                        st.push(c);
                    }
                }
            }
        }
        StringBuilder res = new StringBuilder();
        while (!st.isEmpty()) {
            res.insert(0, st.pop());
        }
        return res.toString();
    }
}
